metadata/src/main/java/org/apache/kafka/timeline/timeline/TimelineHashSet.java

```diff
@@ -0,0 +1,260 @@
+        }
+
+        @Override
+        public boolean hasNext() {
+            return iter.hasNext();
+        }
+
+        @Override
+        public T next() {
+            return iter.next().value;
+        }
+
+        @Override
+        public void remove() {
+            iter.remove();
+        }
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return iterator(Long.MAX_VALUE);
+    }
+
+    public Iterator<T> iterator(long epoch) {
+        return new ValueIterator(epoch);
+    }
+
+    @Override
+    public Object[] toArray() {
+        Object[] result = new Object[size()];
+        Iterator<T> iter = iterator();
+        int i = 0;
+        while (iter.hasNext()) {
+            result[i++] = iter.next();
+        }
+        return result;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <R> R[] toArray(R[] a) {
+        int size = size();
+        if (size <= a.length) {
+            Iterator<T> iter = iterator();
+            int i = 0;
+            while (iter.hasNext()) {
+                a[i++] = (R) iter.next();
+            }
+            while (i < a.length) {
+                a[i++] = null;
+            }
+            return a;
+        } else {
+            return (R[]) toArray();
+        }
+    }
+
+    @Override
+    public boolean add(T newValue) {
+        Objects.requireNonNull(newValue);
+        return snapshottableAddUnlessPresent(new TimelineHashSetEntry<>(newValue));
+    }
+
+    @Override
+    public boolean remove(Object value) {
+        return snapshottableRemove(new TimelineHashSetEntry<>(value)) != null;
+    }
+
+    @Override
+    public boolean containsAll(Collection<?> collection) {
+        for (Object value : collection) {
+            if (!contains(value)) return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends T> collection) {
+        boolean modified = false;
+        for (T value : collection) {
+            if (add(value)) {
+                modified = true;
+            }
+        }
+        return modified;
+    }
+
+    @Override
+    public boolean retainAll(Collection<?> collection) {
+        Objects.requireNonNull(collection);
+        boolean modified = false;
+        Iterator<T> it = iterator();
+        while (it.hasNext()) {
+            if (!collection.contains(it.next())) {
+                it.remove();
+                modified = true;
+            }
+        }
+        return modified;
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> collection) {
+        Objects.requireNonNull(collection);
+        boolean modified = false;
+        Iterator<?> it = iterator();
+        while (it.hasNext()) {
+            if (collection.contains(it.next())) {
+                it.remove();
+                modified = true;
+            }
+        }
+        return modified;
+    }
+
+    @Override
+    public void clear() {
```

Comment:
Should we clear all snapshots in this set too?

Reply:
clear just changes the current state, it doesn't affect snapshotted state